context-extraction/src/main/java/de/atb/context/persistence/context/ContextRepository.java

```diff
@@ -358,21 +278,50 @@ public synchronized Model executeSparqlDescribeQuery(BusinessCase businessCase,
      * @see de.atb.context.persistence.context.IContextRepository#executeSparqlConstructQuery(de.atb.context.common.util.BusinessCase, java.lang.String, boolean)
      */
     @Override
-    public synchronized Model executeSparqlConstructQuery(BusinessCase businessCase, String query, boolean useReasoner) {
+    public synchronized Model executeSparqlConstructQuery(BusinessCase businessCase,
+                                                          String query,
+                                                          boolean useReasoner) {
+        return executeSparql(businessCase, query, useReasoner, QueryExecution::execConstruct);
+    }
+
+    private <R> R executeSparql(BusinessCase businessCase, String query, boolean useReasoner, Function<QueryExecution, R> executioner) {
         if (businessCase == null) {
             throw new NullPointerException("BusinessCase may not be null!");
         }
         Dataset dataset = getDataSet(businessCase);
-        QueryExecution qexec = getQueryExecution(businessCase, query, useReasoner, dataset);
-        Model result = null;
+        // FIXME: Should this be `begin(ReadWrite.READ)`? Because everything in `block` seems to be just "selects"?
+        dataset.begin(ReadWrite.WRITE);
         try {
-            result = qexec.execConstruct();
-        } catch (QueryException qe) {
-            ContextRepository.logger.error(qe.getMessage(), qe);
+            QueryExecution queryExecution = getQueryExecution(businessCase, query, useReasoner, dataset);
+            R result = executioner.apply(queryExecution);
+            dataset.commit();
+            return result;
+        } catch (Exception e) {
+            ContextRepository.logger.error(e.getMessage(), e);
+            dataset.abort();
+            return null;
+        } finally {
+            dataset.end();
         }
-        dataset.commit();
-        dataset.end();
-        return result;
+    }
+
+    private QueryExecution getQueryExecution(BusinessCase businessCase, String query, boolean useReasoner, Dataset dataset) {
+        if (query == null) {
+            throw new NullPointerException("Query may not be null!");
+        }
+        if (query.trim().length() == 0) {
+            throw new IllegalArgumentException("Query may not be empty!");
+        }
+        String finalQuery = prepareSparqlQuery(businessCase, query);
+        dataset.begin(ReadWrite.WRITE);
```

Comment:
This line can be removed because `dataset.begin` has already been called in calling method `executeSparql`.